### PR TITLE
PARQUET-1100: Introduce RecordReader interface to better support nested data, refactor parquet/arrow/reader

### PR DESCRIPTION
**Pull request by wesm**
We did not have very consistent logic around reading values from leaf nodes versus reading semantic records where the repetition level is greater than zero. This introduces a reader class that reads from column chunks until it identifies the end of records. It also reads values (with spaces, if required by the schema) into internal buffers. This permitted a substantial refactoring and simplification of the code in parquet::arrow where we were handling the interpretation of batch reads as records manually. 

As follow up patch, we should be able to take a collection of record readers from the same "tree" in a nested type and reassemble the intermediate Arrow structure and dealing with any redundant structure information in repetition and definition levels. This should a allow a unification of our nested data read code path so that we can read arbitrary nested structures.

### CODE REVIEW

**Comment by wesm** [comment]
Edge cases galore here. We are able to read the file in PARQUET-1100 correctly now:

```python
In [2]: t = pq.read_table('/home/wesm/Downloads/part-00000-6570e34b-b42c-4a39-8adf-21d3a97fb87d.snappy.parquet')

In [3]: t.to
t.to_pandas  t.to_pydict  

In [3]: t.to_pandas().head()
Out[3]: 
   label                                       account_meta  features_type  \
0     21  {'max_arena': 7, 'arena': 7, 'xp_level': 8, 'c...              0   
1     21  {'max_arena': 8, 'arena': 8, 'xp_level': 9, 'c...              0   
2     17  {'max_arena': 7, 'arena': 7, 'xp_level': 7, 'c...              0   
3     22  {'max_arena': 8, 'arena': 8, 'xp_level': 8, 'c...              0   
4     17  {'max_arena': 10, 'arena': 10, 'xp_level': 10,...              0   

   features_size                                   features_indices  \
0           3815  [1, 2, 5, 6, 7, 8, 11, 12, 13, 15, 17, 18, 21,...   
1           3815  [0, 1, 2, 5, 6, 7, 9, 12, 14, 15, 16, 17, 21, ...   
2           3815  [1, 4, 7, 9, 11, 12, 13, 14, 15, 16, 17, 19, 2...   
3           3815  [12, 15, 17, 22, 68, 70, 72, 74, 91, 96, 99, 1...   
4           3815  [1, 5, 8, 15, 17, 21, 24, 41, 68, 79, 85, 89, ...   

                                     features_values  
0  [0.69314718056, 0.69314718056, 3.25809653802, ...  
1  [2.77258872224, 0.69314718056, 1.60943791243, ...  
2  [1.94591014906, 1.38629436112, 2.3978952728, 0...  
3  [1.09861228867, 1.09861228867, 1.09861228867, ...  
4  [0.69314718056, 1.79175946923, 1.79175946923, ...  
```

I'm looking into perf changes now

File: src/parquet/arrow/arrow-reader-writer-test.cc [hunk 1]
```diff
@@ -1287,7 +1336,92 @@ TEST(TestArrowReadWrite, ReadColumnSubset) {
...
+
+  std::unique_ptr<ColumnReader> col_reader;
+  ASSERT_OK(reader->GetColumn(0, &col_reader));
+
+  auto expected = table->column(0)->data()->chunk(0);
+
+  std::vector<std::shared_ptr<Array>> pieces;
+  for (int i = 0; i < num_rows; ++i) {
+    std::shared_ptr<Array> piece;
+    ASSERT_OK(col_reader->NextBatch(1, &piece));
```
Comment (wesm) [inline src/parquet/arrow/arrow-reader-writer-test.cc]: Reading the table one record at a time brought out a lot of edge cases

**Comment by wesm** [comment]
I think this work may have uncovered an issue in the struct reader implementation. Am working through it

**Comment by wesm** [comment]
@itaiin I want to make sure I'm not misunderstanding the test case you wrote. You are creating a schema that looks like

```
optional group-5
  optional group-4
    optional group-3
      optional group-2
        optional group-1
          optional leaf
```

This is being generated with tree depth 5. There was a place where definition levels were being generated from 0 to 4, which isn't right (I think?) because the definition levels here can range from 0 to 6 (where 6 means that the leaf exists and is non null). I fixed the test case by this logic -- I think my refactor only exposed the bug, it was reading correctly "by accident" before.

**Comment by wesm** [comment]
This has performance regressions, so I'll do some digging tomorrow or Friday to fix that

File: src/parquet/column_reader.cc [hunk 2]
```diff
@@ -283,6 +286,405 @@ std::shared_ptr<ColumnReader> ColumnReader::Make(const ColumnDescriptor* descr,
...
+    } else {
+      values_to_read = num_records;
+      records_read = num_records;
+    }
+  }
+  *values_seen = values_to_read;
+  return records_read;
+}
+
+void RecordReader::ReadValuesDense(ColumnReader* reader, int64_t values_to_read) {
```
Comment (wesm) [inline src/parquet/column_reader.cc]: I will need to refactor these so the type-specific hot path can be inlined with no switch statements

**Comment by itaiin** [comment]
[deep_nested_test_fix_patch.txt](https://github.com/apache/parquet-cpp/files/1302551/deep_nested_test_fix_patch.txt)
@wesm regarding the test I believe you are right, the levels should be 0..6 and not 0..4, however its just the test that was wrong it is still reading correctly. 
after fixing that problem, I found two other problems with it:
a) In the optional case, I've neglected to call `RETURN_NOT_OK(child->Accept(this));` in the visitor which caused not actually checking all the way down
b) The test would then fail because i'm checking in the leafs the wrong thing. 
I've attached a patch with a fix for all 3: 
[deep_nested_test_fix_patch.txt](https://github.com/apache/parquet-cpp/files/1302552/deep_nested_test_fix_patch.txt)

**Comment by wesm** [comment]
thanks @itaiin -- I fixed the issues you cited accept for the missing status check. I'll add that

**Comment by itaiin** [comment]
@wesm  np, sorry for the embarrassing mistake ;)

**Comment by wesm** [comment]
I'm working a bunch on this patch, so maybe wait up on code review until I resolve the perf regression

**Comment by wesm** [comment]
This could use some review. 

I'm trying to figure out the perf regression. It seems to be due to an unaligned memcpy someplace that I haven't tracked down yet:

after this patch:

```
$ ./release/arrow-reader-writer-benchmark 
Run on (8 X 4200.16 MHz CPU s)
2017-09-14 11:27:19
Benchmark                                 Time           CPU Iterations
-----------------------------------------------------------------------
BM_ReadColumn<false,Int32Type>      8339054 ns    8339302 ns         84   4.68415GB/s
BM_ReadColumn<true,Int32Type>      45660221 ns   45661955 ns         15   1.28321GB/s
BM_ReadColumn<false,Int64Type>     16125932 ns   16126464 ns         43   4.84452GB/s
BM_ReadColumn<true,Int64Type>      55144648 ns   55146157 ns         12   1.77086GB/s
BM_ReadColumn<false,DoubleType>    16830389 ns   16830884 ns         42   4.64176GB/s
BM_ReadColumn<true,DoubleType>     65370585 ns   65372715 ns         11   1.49384GB/s
```

before:

```
$ ./release/arrow-reader-writer-benchmark 
Run on (8 X 4200.62 MHz CPU s)
2017-09-14 11:28:18
Benchmark                                 Time           CPU Iterations
-----------------------------------------------------------------------
BM_ReadColumn<false,Int32Type>      3107818 ns    3107899 ns        225   12.5688GB/s
BM_ReadColumn<true,Int32Type>      32462371 ns   32463373 ns         21   1.80492GB/s
BM_ReadColumn<false,Int64Type>      5801545 ns    5801697 ns        121   13.4659GB/s
BM_ReadColumn<true,Int64Type>      37961699 ns   37962768 ns         18   2.57242GB/s
BM_ReadColumn<false,DoubleType>     6849948 ns    6850132 ns        102   11.4049GB/s
BM_ReadColumn<true,DoubleType>     46397502 ns   46398727 ns         15   2.10472GB/s
```

On these simple flat column read paths, I don't really see a reason for the regression, but running FlameGraph suggests it's due to memory alignment.

There's a bit of code duplication in RecordReader vs. ColumnReader. Ideally ColumnReader would go away eventually because RecordReader is what we want for general nested data support, but I left the existing code as is for backwards compatibility.

**Comment by wesm** [comment]
One of the perf regressions was due to failure to preallocate, that's fixed now:

```
$ ./release/arrow-reader-writer-benchmark 
Run on (8 X 4200.31 MHz CPU s)
2017-09-14 11:42:05
Benchmark                                 Time           CPU Iterations
-----------------------------------------------------------------------
BM_ReadColumn<false,Int32Type>      2885395 ns    2885473 ns        241   13.5376GB/s
BM_ReadColumn<true,Int32Type>      43588320 ns   43589747 ns         16   1.34421GB/s
BM_ReadColumn<false,Int64Type>      5598802 ns    5598965 ns        125   13.9535GB/s
BM_ReadColumn<true,Int64Type>      43625578 ns   43627261 ns         16   2.23842GB/s
BM_ReadColumn<false,DoubleType>     6667011 ns    6667135 ns        105   11.7179GB/s
BM_ReadColumn<true,DoubleType>     54040599 ns   54042125 ns         13   1.80704GB/s
```

It seems that nullable reads are 10-20% slower. I will investigate that some more, but this can be reviewed in the meantime

**Comment by wesm** [comment]
Alright, I figured out the perf regression, I think. The reason is that we are looking at the definition levels twice (once to delimit records, then again to populate the null bitmap) -- in the case of flat data we can skip the record delimiting logic. Will take me a little hacking to fix but then I think we should be back to original performance. There's also one valgrind warning I need to investigate.

**Comment by itaiin** [comment]
Cool finding. 
I have found a bug with a valgrind warning in TypedReadBatch for boolean type literally half an hour ago: first call to `memcpy` (arrow/reader.cc:1142) runs over some memory, I think its because the `Resize` in the prev. line uses a wrong size... might this be it?

**Comment by wesm** [comment]
OK, perf regression is fixed. I will look into the valgrind issue shortly, thank you for the pointer!

```
$ ./release/arrow-reader-writer-benchmark 
Run on (8 X 4200.16 MHz CPU s)
2017-09-14 13:40:59
Benchmark                                 Time           CPU Iterations
-----------------------------------------------------------------------
BM_ReadColumn<false,Int32Type>      2967900 ns    2967969 ns        236   13.1614GB/s
BM_ReadColumn<true,Int32Type>      33400825 ns   33402087 ns         21   1.75419GB/s
BM_ReadColumn<false,Int64Type>      5696175 ns    5696390 ns        123   13.7148GB/s
BM_ReadColumn<true,Int64Type>      38892670 ns   38893906 ns         18   2.51084GB/s
BM_ReadColumn<false,DoubleType>     6745839 ns    6746094 ns        104   11.5808GB/s
BM_ReadColumn<true,DoubleType>     42422320 ns   42423468 ns         17   2.30194GB/s
```

**Comment by wesm** [comment]
@itaiin arrow/reader.cc has been half-rewritten in this patch. Do you still see that issue in this branch? I am looking into the last valgrind warning in arrow-reader-writer-test that's happening here

**Comment by itaiin** [comment]
@wesm sorry didn't realize the change was this big.. seems from the diff like the faulty code was removed.

**Comment by wesm** [comment]
OK, I'm done with this patch modulo code review! I'll work on getting the build back to green

**Comment by wesm** [comment]
I removed `RecordReader` from the public API since it's really an implementation detail of reading from Parquet->Arrow.

File: src/parquet/arrow/reader.cc [hunk 3]
```diff
@@ -428,24 +358,26 @@ Status FileReader::Impl::ReadSchemaField(int i, const std::vector<int>& indices,
 
   std::unique_ptr<ColumnReader> reader(new ColumnReader(std::move(reader_impl)));
 
-  int64_t batch_size = 0;
+  // TODO(wesm): This calculation doesn't make much sense when we have repeated
+  // schema nodes
+  int64_t records_to_read = 0;
   for (int j = 0; j < reader_->metadata()->num_row_groups(); j++) {
-    batch_size += reader_->metadata()->RowGroup(j)->ColumnChunk(i)->num_values();
+    records_to_read += reader_->metadata()->RowGroup(j)->ColumnChunk(i)->num_values();
```
Comment (cpcloud) [inline src/parquet/arrow/reader.cc]: Does it make sense to bring `reader_->metadata()` outside of the loop to reduce the number of pointer dereferences?

File: src/parquet/arrow/reader.cc [hunk 4]
```diff
@@ -1032,346 +695,291 @@ Status PrimitiveImpl::WrapIntoListArray(const int16_t* def_levels,
...
+                    std::shared_ptr<Array>* out) {
+    int64_t length = reader->values_written();
+    auto values = reinterpret_cast<const int32_t*>(reader->values());
+
+    std::shared_ptr<Buffer> data;
+    RETURN_NOT_OK(::arrow::AllocateBuffer(pool, length * sizeof(int64_t), &data));
+    auto out_ptr = reinterpret_cast<int64_t*>(data->mutable_data());
+
+    for (int64_t i = 0; i < length; i++) {
+      *out_ptr++ = static_cast<int64_t>(values[i]) * 86400000;
```
Comment (cpcloud) [inline src/parquet/arrow/reader.cc]: Should the `86400000` be a global constant?

File: src/parquet/arrow/reader.cc [hunk 5]
```diff
@@ -1467,16 +1075,21 @@ Status StructImpl::NextBatch(int batch_size, std::shared_ptr<Array>* out) {
...
   }
 
   RETURN_NOT_OK(DefLevelsToNullArray(&null_bitmap, &null_count));
 
-  *out = std::make_shared<StructArray>(field()->type(), batch_size, children_arrays,
-                                       null_bitmap, null_count);
+  int64_t struct_length = children_arrays[0]->length();
+  for (size_t i = 1; i < children_arrays.size(); ++i) {
+    if (children_arrays[i]->length() != struct_length) {
+      return Status::Invalid("Struct children had different lengths");
```
Comment (cpcloud) [inline src/parquet/arrow/reader.cc]: Should this be a `DCHECK`? Or is it possible for this to happen because of error on the user's part?

File: src/parquet/arrow/record_reader.cc [hunk 6]
```diff
@@ -0,0 +1,804 @@
...
+
+template <typename DType>
+class TypedRecordReader : public RecordReader::RecordReaderImpl {
+ public:
+  typedef typename DType::c_type T;
+
+  ~TypedRecordReader() {}
+
+  TypedRecordReader(const ColumnDescriptor* schema, ::arrow::MemoryPool* pool)
+      : RecordReader::RecordReaderImpl(schema, pool), current_decoder_(NULL) {}
```
Comment (cpcloud) [inline src/parquet/arrow/record_reader.cc]: Should `NULL` be `nullptr`?

File: src/parquet/arrow/record_reader.cc [hunk 7]
```diff
@@ -0,0 +1,804 @@
...
+    if (levels_position_ < levels_written_) {
+      records_read += ReadRecordData(num_records);
+    }
+
+    // HasNext invokes ReadNewPage
+    if (records_read == 0 && !HasNext()) {
+      return 0;
+    }
+
+    constexpr int64_t kLevelBatchSize = 1024;
```
Comment (cpcloud) [inline src/parquet/arrow/record_reader.cc]: Should this be a global constant?

File: src/parquet/arrow/record_reader.cc [hunk 8]
```diff
@@ -0,0 +1,804 @@
...
+          new RecordReader(new TypedRecordReader<FloatType>(descr, pool)));
+    case Type::DOUBLE:
+      return std::shared_ptr<RecordReader>(
+          new RecordReader(new TypedRecordReader<DoubleType>(descr, pool)));
+    case Type::BYTE_ARRAY:
+      return std::shared_ptr<RecordReader>(
+          new RecordReader(new TypedRecordReader<ByteArrayType>(descr, pool)));
+    case Type::FIXED_LEN_BYTE_ARRAY:
+      return std::shared_ptr<RecordReader>(
+          new RecordReader(new TypedRecordReader<FLBAType>(descr, pool)));
```
[the remark this answers is not in the document]
Comment (wesm) [reply src/parquet/arrow/record_reader.cc]: Private constructor

[the remark this answers is not in the document]
Comment (cpcloud) [reply src/parquet/arrow/record_reader.cc]: Ah, right.

**Comment by itaiin** [comment]
I'm not 100% clear about the motivation so a question about the design (while risking barging into an open door :)) - Looking at the current code, is the assumption that the reader tree stays with `ColumnReader` while only the primitives/leaf nodes use `RecordReader`?

The motivation to the question is that `ColumnReader::NextBatch` is fairly limiting for constructing a reader tree for kind of the same reason `RecordReader` was born, since repeating values can happen at any level of the tree (i.e. a list of structs). 
I imagine something like reading from a lower level `RecordReader` and constructing values into the current level, and only finalizing the builders to arrays at the top level into a fully fledged array.

Seems to me that one thing that currently limits this is the dependency on `ColumnDescriptor`, which can be used only for primitives, although its parameters - the schema node, and the max levels are relevant also to nested types. (perhaps there are more but I can't seem to think of any right now).

**Comment by wesm** [comment]
@itaiin this patch deliberately makes no progress toward general record reassembly but deals with the problem of needing to delimit repeated records, which wasn't being handled properly before this. I anticipate significantly more refactoring as part of general nested data reassembly, but either way we need a stateful record reader class that can scan until identifying the

**Comment by itaiin** [comment]
@wesm ok cool.
After this PR is done, i will try to merge with my full nested reader efforts and make the discussion more concrete.

File: src/parquet/arrow/arrow-reader-writer-benchmark.cc [hunk 9]
```diff
@@ -29,12 +31,13 @@
 
 using arrow::NumericBuilder;
 
-#define ABORT_NOT_OK(s)                  \
-  do {                                   \
-    ::arrow::Status _s = (s);            \
-    if (ARROW_PREDICT_FALSE(!_s.ok())) { \
-      exit(-1);                          \
-    }                                    \
+#define ABORT_NOT_OK(s)                                       \
```
[the remark this answers is not in the document]
Comment (wesm) [reply src/parquet/arrow/arrow-reader-writer-benchmark.cc]: done

File: src/parquet/arrow/arrow-reader-writer-benchmark.cc [hunk 10]
```diff
@@ -29,12 +31,13 @@
...
+  do {                                                        \
+    ::arrow::Status _s = (s);                                 \
+    if (ARROW_PREDICT_FALSE(!_s.ok())) {                      \
+      std::cout << "Exiting: " << _s.ToString() << std::endl; \
+      exit(-1);                                               \
```
[the remark this answers is not in the document]
Comment (wesm) [reply src/parquet/arrow/arrow-reader-writer-benchmark.cc]: done

File: src/parquet/column_reader.h [hunk 11]
```diff
@@ -130,7 +141,61 @@ class PARQUET_EXPORT ColumnReader {
...
-// API to read values from a single column. This is the main client facing API.
+namespace internal {
+
+static inline void DefinitionLevelsToBitmap(const int16_t* def_levels,
+                                            const int64_t num_def_levels,
+                                            const int16_t max_definition_level,
+                                            const int16_t max_repetition_level,
+                                            int64_t* values_read, int64_t* null_count,
+                                            uint8_t* valid_bits,
+                                            const int64_t valid_bits_offset) {
```
[the remark this answers is not in the document]
Comment (wesm) [reply src/parquet/column_reader.h]: Do you know if C/C++ compilers will generate better code if parameters are marked const? If they can be analyzed as invariant then perhaps there is no difference

File: src/parquet/column_reader.h [hunk 12]
```diff
@@ -130,7 +141,61 @@ class PARQUET_EXPORT ColumnReader {
...
+      if (def_levels[i] < max_definition_level) {
+        bitset &= ~(1 << bit_offset);
+        *null_count += 1;
+      } else {
+        throw ParquetException("definition level exceeds maximum");
+      }
+    }
+
+    bit_offset++;
+    if (bit_offset == 8) {
```
[the remark this answers is not in the document]
Comment (wesm) [reply src/parquet/column_reader.h]: done. though I suppose if `CHAR_BIT` is not 8 then I wouldn't expect this library to work

File: src/parquet/column_reader.h [hunk 13]
```diff
@@ -130,7 +141,61 @@ class PARQUET_EXPORT ColumnReader {
...
+      valid_bits[byte_offset] = bitset;
+      byte_offset++;
+      // TODO: Except for the last byte, this shouldn't be needed
+      bitset = valid_bits[byte_offset];
+    }
+  }
+  if (bit_offset != 0) {
+    valid_bits[byte_offset] = bitset;
+  }
+  *values_read = (bit_offset + byte_offset * 8 - valid_bits_offset);
```
[the remark this answers is not in the document]
Comment (wesm) [reply src/parquet/column_reader.h]: fixed

File: src/parquet/column_reader.h [hunk 14]
```diff
@@ -294,59 +363,38 @@ inline int64_t TypedColumnReader<DType>::ReadBatch(int64_t batch_size,
...
-      // non-repeated+nested case
-      if (def_levels[i] < max_definition_level) {
-        bitset &= ~(1 << bit_offset);
-        *null_count += 1;
-      } else {
-        throw ParquetException("definition level exceeds maximum");
+// TODO(itaiin): another code path split to merge when the general case is done
+static inline bool HasSpacedValues(const ColumnDescriptor* descr) {
+  // True if the node is nullable, or if it's contained in a nullable group
+  // prior to the previous repeated node
```
[the remark this answers is not in the document]
Comment (wesm) [reply src/parquet/column_reader.h]: I removed this comment; it was outdated. we'll need to expand this logic when dealing with complete nested data reads

File: src/parquet/column_reader.h [hunk 15]
```diff
@@ -294,59 +363,38 @@ inline int64_t TypedColumnReader<DType>::ReadBatch(int64_t batch_size,
...
+    return !descr->schema_node()->is_required();
+  } else {
+    // non-repeated+nested case
+    // Find if a node forces nulls in the lowest level along the hierarchy
+    const schema::Node* node = descr->schema_node().get();
+    while (node) {
+      auto parent = node->parent();
+      if (node->is_optional()) {
+        return true;
+        break;
```
[the remark this answers is not in the document]
Comment (wesm) [reply src/parquet/column_reader.h]: fixed

File: src/parquet/column_reader.h [hunk 16]
```diff
@@ -294,59 +363,38 @@ inline int64_t TypedColumnReader<DType>::ReadBatch(int64_t batch_size,
...
       }
+      node = parent;
```
[the remark this answers is not in the document]
Comment (wesm) [reply src/parquet/column_reader.h]: fixed (this was copy-pasted)

File: src/parquet/column_reader.h [hunk 17]
```diff
@@ -377,25 +425,7 @@ inline int64_t TypedColumnReader<DType>::ReadBatchSpaced(
...
-      while (node) {
-        auto parent = node->parent();
-        if (node->is_optional()) {
-          has_spaced_values = true;
-          break;
-        }
-        node = parent;
-      }
-    }
+    bool has_spaced_values = internal::HasSpacedValues(descr_);
```
[the remark this answers is not in the document]
Comment (wesm) [reply src/parquet/column_reader.h]: fixed

File: src/parquet/arrow/record_reader.cc (hunk 6 above)
Comment (wesm) [reply src/parquet/arrow/record_reader.cc]: fixed. i went through and replaced other usages of NULL in the codebase with nullptr

File: src/parquet/arrow/record_reader.cc (hunk 7 above)
Comment (wesm) [reply src/parquet/arrow/record_reader.cc]: done. also added comment explaining what it is

File: src/parquet/arrow/reader.cc (hunk 4 above)
Comment (wesm) [reply src/parquet/arrow/reader.cc]: done

File: src/parquet/arrow/reader.cc (hunk 5 above)
Comment (wesm) [reply src/parquet/arrow/reader.cc]: This would occur if the Parquet file is malformed somehow. I'll leave a TODO to consider this

File: src/parquet/arrow/arrow-reader-writer-test.cc [hunk 18]
```diff
@@ -344,6 +347,45 @@ static std::shared_ptr<GroupNode> MakeSimpleSchema(const ::arrow::DataType& type
   return std::static_pointer_cast<GroupNode>(node_);
 }
 
+void AssertArraysEqual(const Array& expected, const Array& actual) {
+  if (!actual.Equals(expected)) {
```
[the remark this answers is not in the document]
Comment (wesm) [reply src/parquet/arrow/arrow-reader-writer-test.cc]: sounds good. Created ARROW-1556

File: src/parquet/arrow/reader.cc [hunk 19]
```diff
@@ -223,96 +195,54 @@ class FileReader::Impl {
   int num_threads_;
 };
 
-typedef const int16_t* ValueLevelsPtr;
-
-class ColumnReader::Impl {
+class ColumnReader::ColumnReaderImpl {
```
[the remark this answers is not in the document]
Comment (wesm) [reply src/parquet/arrow/reader.cc]: I tried to do this, but the compiler doesn't like it because the PIMPL classes were originally declared in an anonymous namespace

File: src/parquet/arrow/reader.cc [hunk 20]
```diff
@@ -193,8 +164,9 @@ class FileReader::Impl {
   Status ReadSchemaField(int i, std::shared_ptr<Array>* out);
   Status ReadSchemaField(int i, const std::vector<int>& indices,
                          std::shared_ptr<Array>* out);
-  Status GetReaderForNode(int index, const NodePtr& node, const std::vector<int>& indices,
-                          int16_t def_level, std::unique_ptr<ColumnReader::Impl>* out);
+  Status GetReaderForNode(int index, const Node* node, const std::vector<int>& indices,
```
[the remark this answers is not in the document]
Comment (wesm) [reply src/parquet/arrow/reader.cc]: done

**Comment by wesm** [comment]
OK, this is rebased and ready to go as soon as the build is passing again

File: src/parquet/column_reader.h (hunk 11 above)
[the remark this answers is not in the document]
Comment (cpcloud) [reply src/parquet/column_reader.h]: Actually, after reading around I think it's probably *better* to declare these `const`. You can't modify the value (local to the function of course) which makes reasoning about it easier. So, this can be left as is.

**Comment by wesm** [comment]
Seems like Travis is having a rough day. Here's a Travis CI build going now on my fork, I will merge once all the builds run: https://travis-ci.org/wesm/parquet-cpp/builds/277416117
